client/allocrunner/checks_hook.go

```diff
@@ -0,0 +1,270 @@
+
+			// create the observer for this check
+			h.observers[id] = &observer{
+				ctx:        ctx,
+				cancel:     cancel,
+				check:      check.Copy(),
+				checkStore: h.shim,
+				checker:    h.checker,
+				allocID:    h.allocID,
+				qc: &checks.QueryContext{
+					ID:               id,
+					CustomAddress:    service.Address,
+					ServicePortLabel: service.PortLabel,
+					Ports:            ports,
+					Networks:         networks,
+					NetworkStatus:    h.network,
+					Group:            alloc.Name,
+					Task:             service.TaskName,
+					Service:          service.Name,
+					Check:            check.Name,
+				},
+			}
+
+			// insert a pending result into state store for each check
+			result := checks.Stub(id, structs.GetCheckMode(check), now, alloc.Name, service.TaskName, service.Name, check.Name)
+			if err := h.shim.Set(h.allocID, result); err != nil {
+				h.logger.Error("failed to set initial check status", "id", h.allocID, "error", err)
+				continue
+			}
+
+			// start the observer
+			go h.observers[id].start()
+		}
+	}
+}
+
+func (h *checksHook) Name() string {
+	return checksHookName
+}
+
+func (h *checksHook) Prerun() error {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	group := h.alloc.Job.LookupTaskGroup(h.alloc.TaskGroup)
+	if group == nil {
+		return nil
+	}
+
+	// create and start observers of nomad service checks in alloc
+	h.observe(h.alloc, group.NomadServices())
+
+	return nil
+}
+
+func (h *checksHook) Update(request *interfaces.RunnerUpdateRequest) error {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	group := request.Alloc.Job.LookupTaskGroup(request.Alloc.TaskGroup)
+	if group == nil {
+		return nil
+	}
+
+	// get all group and task level services using nomad provider
+	services := group.NomadServices()
+
+	// create a set of the updated set of checks
+	next := make([]structs.CheckID, 0, len(h.observers))
+	for _, service := range services {
+		for _, check := range service.Checks {
+			next = append(next, structs.NomadCheckID(
+				request.Alloc.ID,
+				request.Alloc.TaskGroup,
+				check,
+			))
+		}
+	}
+
+	// stop the observers of the checks we are removing
+	remove := h.shim.Difference(request.Alloc.ID, next)
+	for _, id := range remove {
+		h.observers[id].stop()
+		delete(h.observers, id)
+	}
+
+	// remove checks that are no longer part of the allocation
+	if err := h.shim.Remove(request.Alloc.ID, remove); err != nil {
+		return err
+	}
+
+	// remember this new alloc
+	h.alloc = request.Alloc
+
+	// ensure we are observing new checks (idempotent)
+	h.observe(request.Alloc, services)
+
+	return nil
+}
+
+func (h *checksHook) PreKill() {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	// terminate our hook context, which threads down to all observers
+	h.stop()
+
+	// purge all checks for this allocation from the client state store
+	if err := h.shim.Purge(h.allocID); err != nil {
+		h.logger.Error("failed to purge check results", "alloc_id", h.allocID, "error", err)
+	}
```

Comment:
Not really, no. PreKill doesn't report an error either so it's not like we can prevent the client from continuing with purging the alloc - though presumably if the state store can't remove a check, it can't remove anything else, either